src/physics/base/PhysicsStepUtils.i.hh

```diff
@@ -144,74 +169,91 @@ CELER_FUNCTION ParticleTrackView::Energy
-    real_type eloss = total_eloss_rate * step;
-
-    if (eloss > pre_step_energy.value() * physics.linear_loss_limit())
+    if (eloss >= pre_step_energy * physics.linear_loss_limit())
     {
         // Enough energy is lost over this step that the dE/dx linear
         // approximation is probably wrong. Use the definition of the range as
         // the integral of 1/loss to back-calculate the actual energy loss
         // along the curve given the actual step.
-        eloss = 0;
-        if (auto ppid = physics.eloss_ppid())
-        {
-            if (auto grid_id = physics.value_grid(VGT::range, ppid))
-            {
-                // Recalculate beginning-of-step range (instead of storing)
-                auto calc_range
-                    = physics.make_calculator<RangeCalculator>(grid_id);
-                real_type remaining_range = calc_range(pre_step_energy) - step;
-                CELER_ASSERT(remaining_range >= 0);
+        auto grid_id = physics.value_grid(VGT::range, ppid);
+        CELER_ASSERT(grid_id);
 
-                // Calculate energy along the range curve corresponding to the
-                // actual step taken: this gives the exact energy loss over the
-                // step due to this process.
-                auto calc_energy
-                    = physics.make_calculator<InverseRangeCalculator>(grid_id);
-                eloss = (pre_step_energy.value()
-                         - calc_energy(remaining_range).value());
-            }
+        // Recalculate beginning-of-step range (instead of storing)
+        auto calc_range = physics.make_calculator<RangeCalculator>(grid_id);
+        real_type range = calc_range(Energy{pre_step_energy});
+        if (step == range)
+        {
+            // TODO: eloss should be pre_step_energy if and only if the range
+            // was the  step limiter (step == range). When we refactor to have
+            // generic range limiters, this exception should no longer be
+            // needed.
+            return Energy{pre_step_energy};
         }
-        CELER_ASSERT(eloss > 0);
-        CELER_ASSERT(eloss <= pre_step_energy.value());
+        CELER_ASSERT(range > step);
+
+        // Calculate energy along the range curve corresponding to the actual
+        // step taken: this gives the exact energy loss over the step due to
+        // this process. If the step is very near the range (a few ULP off, for
+        // example), then the post-step energy will be calculated as zero
+        // without going through the condition above.
+        auto calc_energy
+            = physics.make_calculator<InverseRangeCalculator>(grid_id);
+        eloss = pre_step_energy - value_as<Energy>(calc_energy(range - step));
     }
 
-    // Add energy loss fluctuations if this is the "energy loss" process
-    if (eloss > 0 && eloss < pre_step_energy.value()
-        && physics.add_fluctuation())
+    if (physics.add_fluctuation() && eloss > 0 && eloss < pre_step_energy)
```

Comment:
I had the same thought -- which is why the last commit is there, because that led to an infinite loop. What was happening is that `range - step` was someting like 1e-30, which meant that `eloss = pre_step_energy`. I think we can change it to an assertion once we have different "step limiters", but for now it has to be in this conditional.